flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/AlternatingController.java

```diff
@@ -74,82 +90,152 @@ public void barrierAnnouncement(
+        activeController.barrierReceived(
+                channelInfo,
+                barrier,
+                checkpointBarrier -> {
+                    throw new IllegalStateException("Control should not trigger a checkpoint");
+                });
 
         if (maybeTimedOut.isPresent()) {
             if (activeController == alignedController) {
-                switchToUnaligned(channelInfo, maybeTimedOut.get());
-                return maybeTimedOut;
+                switchToUnaligned(channelInfo, maybeTimedOut.get(), b -> {});
+                triggerCheckpoint.accept(maybeTimedOut.get());
             } else {
                 alignedController.resumeConsumption(channelInfo);
             }
         } else if (!barrier.getCheckpointOptions().isUnalignedCheckpoint()
                 && activeController == unalignedController) {
             alignedController.resumeConsumption(channelInfo);
         }
-        return Optional.empty();
     }
 
     @Override
-    public Optional<CheckpointBarrier> preProcessFirstBarrier(
-            InputChannelInfo channelInfo, CheckpointBarrier barrier)
+    public void preProcessFirstBarrier(
+            InputChannelInfo channelInfo,
+            CheckpointBarrier barrier,
+            ThrowingConsumer<CheckpointBarrier, IOException> triggerCheckpoint)
             throws IOException, CheckpointException {
         if (lastSeenBarrier < barrier.getId()) {
             lastSeenBarrier = barrier.getId();
-            firstBarrierArrivalTime = getArrivalTime(barrier);
+            lastBarrierArrivalTime = getArrivalTime(barrier);
+            if (barrier.getCheckpointOptions().isTimeoutable()
+                    && activeController == alignedController) {
+                scheduleSwitchToUnaligned(channelInfo, barrier, triggerCheckpoint);
+            }
         }
         activeController = chooseController(barrier);
-        return activeController.preProcessFirstBarrier(channelInfo, barrier);
+        activeController.preProcessFirstBarrier(channelInfo, barrier, triggerCheckpoint);
     }
 
-    private void switchToUnaligned(InputChannelInfo channelInfo, CheckpointBarrier barrier)
+    private void scheduleSwitchToUnaligned(
+            InputChannelInfo channelInfo,
+            CheckpointBarrier barrier,
+            ThrowingConsumer<CheckpointBarrier, IOException> triggerCheckpoint) {
+        delayedActionRegistration.schedule(
+                () -> {
+                    long barrierId = barrier.getId();
+                    if (lastSeenBarrier == barrierId
+                            && lastCompletedBarrier < barrierId
+                            && activeController == alignedController) {
+                        switchToUnaligned(channelInfo, barrier.asUnaligned(), triggerCheckpoint);
+                    }
+                    return null;
+                },
+                Duration.ofMillis(barrier.getCheckpointOptions().getAlignmentTimeout() + 1));
```

Comment:
The +1 is because of what we considered timed out before:
```
barrier.getCheckpointOptions().getAlignmentTimeout() * 1_000_000
                        < (System.nanoTime() - firstBarrierArrivalTime)
```
firstBarrierArrivalTime + alignment timeout was not considered as timed out.

Reply:
I see, thanks.